File: 0804/LibrarySystem_Python/library.py
import json

from book import Book



class Library:


    def __init__(self):

        self.books = []



    # 도서 등록

    def add_book(
        self,
        title,
        author,
        publisher
    ):


        # 중복 확인

        for book in self.books:

            if book.title == title:

                return False



        book = Book(

            title,

            author,

            publisher,

            False

        )


        self.books.append(book)


        return True





    # 전체 도서 반환

    def get_books(self):

        return self.books






    # 검색

    def search_book(
        self,
        keyword
    ):


        result = []


        for book in self.books:


            if keyword.lower() in book.title.lower():


                result.append(book)



        return result






    # 삭제

    def delete_book(
        self,
        title
    ):


        for book in self.books:


            if book.title == title:


                self.books.remove(book)


                return True



        return False






    # 대출

    def borrow_book(
        self,
        title
    ):


        for book in self.books:


            if book.title == title:


                if book.borrowed:


                    return False



                book.borrowed = True


                return True




        return False






    # 반납

    def return_book(
        self,
        title
    ):


        for book in self.books:


            if book.title == title:


                if not book.borrowed:


                    return False



                book.borrowed = False


                return True




        return False






    # 저장

    def save_file(self):


        data = []



        for book in self.books:


            data.append(

                book.to_dict()

            )




        with open(

            "books.json",

            "w",

            encoding="utf-8"

        ) as file:


            json.dump(

                data,

                file,

                ensure_ascii=False,

                indent=4

            )



        return True






    # 불러오기

    def load_file(self):


        try:


            with open(

                "books.json",

                "r",

                encoding="utf-8"

            ) as file:


                data = json.load(file)





            self.books.clear()



            for item in data:


                self.books.append(

                    Book.from_dict(item)

                )



            return True





        except FileNotFoundError:


            return False

File: 0804/LibrarySystem/LibrarySystem/Library.cpp
#include "Library.h"
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;

// 도서 등록
void Library::addBook()
{
    string title;
    string author;
    string publisher;

    cout << "\n=== 도서 등록 ===\n";

    cin.ignore();

    cout << "제목 : ";
    getline(cin, title);

    cout << "저자 : ";
    getline(cin, author);

    cout << "출판사 : ";
    getline(cin, publisher);

    Book book(title, author, publisher);

    books.push_back(book);

    cout << "도서가 등록되었습니다.\n";
}

void Library::saveFile()
{
    ofstream fout("books.txt");

    if (!fout)
    {
        cout << "파일을 열 수 없습니다.\n";
        return;
    }

    for (int i = 0; i < books.size(); i++)
    {
        fout << books[i].saveData() << endl;
    }

    fout.close();

    cout << "파일 저장 완료!\n";
}

void Library::loadFile()
{
    ifstream fin("books.txt");

    if (!fin)
    {
        cout << "저장된 파일이 없습니다.\n";
        return;
    }

    books.clear();

    string line;

    while (getline(fin, line))
    {
        stringstream ss(line);

        string title;
        string author;
        string publisher;
        string borrowedStr;

        getline(ss, title, '|');
        getline(ss, author, '|');
        getline(ss, publisher, '|');
        getline(ss, borrowedStr);

        bool borrowed = (borrowedStr == "1");

        Book book(title, author, publisher, borrowed);

        books.push_back(book);
    }

    fin.close();

    cout << "파일 불러오기 완료!\n";
}

// 전체 목록
void Library::showBooks()
{
    cout << "\n=== 도서 목록 ===\n";

    if (books.empty())
    {
        cout << "등록된 도서가 없습니다.\n";
        return;
    }

    for (int i = 0; i < books.size(); i++)
    {
        cout << "[" << i + 1 << "]\n";
        books[i].printBook();
    }
}

// 검색
void Library::searchBook()
{
    string keyword;

    cout << "\n검색할 제목 : ";

    cin.ignore();

    getline(cin, keyword);

    bool found = false;

    for (int i = 0; i < books.size(); i++)
    {
        if (books[i].getTitle() == keyword)
        {
            books[i].printBook();
            found = true;
        }
    }

    if (!found)
    {
        cout << "검색 결과가 없습니다.\n";
    }
}

void Library::borrowBook()
{
    string title;

    cout << "\n=== 도서 대출 ===\n";

    cin.ignore();

    cout << "대출할 도서 제목 : ";
    getline(cin, title);

    for (int i = 0; i < books.size(); i++)
    {
        if (books[i].getTitle() == title)
        {
            if (books[i].isBorrowed())
            {
                cout << "이미 대출 중인 도서입니다.\n";
            }
            else
            {
                books[i].borrowBook();
                cout << "대출이 완료되었습니다.\n";
            }

            return;
        }
    }

    cout << "해당 도서를 찾을 수 없습니다.\n";
}

void Library::returnBook()
{
    string title;

    cout << "\n=== 도서 반납 ===\n";

    cin.ignore();

    cout << "반납할 도서 제목 : ";
    getline(cin, title);

    for (int i = 0; i < books.size(); i++)
    {
        if (books[i].getTitle() == title)
        {
            if (!books[i].isBorrowed())
            {
                cout << "이미 반납된 도서입니다.\n";
            }
            else
            {
                books[i].returnBook();
                cout << "반납이 완료되었습니다.\n";
            }

            return;
        }
    }

    cout << "해당 도서를 찾을 수 없습니다.\n";
}

void Library::deleteBook()
{
    string title;

    cout << "\n=== 도서 삭제 ===\n";

    cin.ignore();

    cout << "삭제할 도서 제목 : ";
    getline(cin, title);

    for (int i = 0; i < books.size(); i++)
    {
        if (books[i].getTitle() == title)
        {
            books.erase(books.begin() + i);

            cout << "도서가 삭제되었습니다.\n";

            return;
        }
    }

    cout << "해당 도서를 찾을 수 없습니다.\n";
}